jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java

```diff
@@ -71,42 +90,47 @@ DbMetrics getDbMetrics() {
     return dbManager.getMetrics();
   }
 
-  private <T> T query(IQuery<DB, T> query, boolean asTransaction) {
-    DB connection = dbManager.getConnection();
-    boolean connectionSafeToReturn = true;
-
-    try {
-      connection.setAutoCommit(!asTransaction);
-      long startTime = System.currentTimeMillis();
-      T value = query.query(connection);
-      if (asTransaction) {
-        connection.commit();
-      }
-      long executionTime = System.currentTimeMillis() - startTime;
-      if (metricsTrackingEnabled) {
-        queryMetrics.update(executionTime, Thread.currentThread().getStackTrace()[3]);
-      }
-
-      return value;
-    } catch (Exception e) {
-      LOG.error("SQL execution failure", e);
-      if (asTransaction) {
-        connectionSafeToReturn = tryToSafelyRollback(connection);
-      }
-
-      throw new SqlExecutionFailureException(e);
-    } catch (Throwable t) {
-      // We still try to explicitly rollback the transaction if a throwable is thrown.
-      if (asTransaction) {
-        connectionSafeToReturn = tryToSafelyRollback(connection);
-      }
-
-      throw t;
-    } finally {
-      if (connectionSafeToReturn) {
-        dbManager.returnConnection(connection);
-      } else {
-        dbManager.invalidateConnection(connection);
+  private <T> T query(IQuery<DB, T> query, ExecutionContext context) {
+    while (true) {
+      DB connection = dbManager.getConnection();
+      boolean connectionSafeToReturn = true;
+      try {
+        connection.setAutoCommit(!context.asTransaction);
+        long startTime = System.currentTimeMillis();
+        T value = query.query(connection);
+        if (context.asTransaction) {
+          connection.commit();
+        }
+        long executionTime = System.currentTimeMillis() - startTime;
+        context.retryPolicy.updateOnSuccess();
+        if (metricsTrackingEnabled) {
+          queryMetrics.update(executionTime, Thread.currentThread().getStackTrace()[3]);
+        }
+        return value;
+      } catch (Exception e) {
+        LOG.error("SQL execution failure", e);
+        if (context.asTransaction) {
+          connectionSafeToReturn = tryToSafelyRollback(connection);
+        }
+
+        context.retryPolicy.updateOnFailure();
+        if (!context.retryPolicy.shouldRetry()) {
+          throw new SqlExecutionFailureException(e);
+        }
+      } catch (Throwable t) {
+        // We still try to explicitly rollback the transaction if a throwable is thrown.
+        if (context.asTransaction) {
+          connectionSafeToReturn = tryToSafelyRollback(connection);
+        }
+
+        throw t;
+      } finally {
+        if (connectionSafeToReturn) {
+          dbManager.returnConnection(connection);
+        } else {
+          dbManager.invalidateConnection(connection);
+        }
+        context.retryPolicy.execute();
```

Comment:
nit: comment explaining why we want to retry after returning/invalidating connections

Reply:
Will do!